### PR TITLE
[Core][CUDA Graph] add output buffer for cudagraph to reduce memory footprint

### PR DESCRIPTION
**Pull request by youkaichao**
Currently, each cudagraph for one batch size will incur additional memory cost of `TP * BATCH_SIZE * HIDDEN_SIZE` . Then the total memory footprint for all cudagraphs are `TP * SUM(BATCH_SIZES) * HIDDEN_SIZE`.

If we create output buffer too, the total memory footprint for all cudagraphs are `TP * MAX(BATCH_SIZES) * HIDDEN_SIZE`, which is constant, regardless of how many graphs we capture.

### CODE REVIEW

**Comment by youkaichao** [comment]
Performance-wise, this adds one memory copy operation into cudagraph. This should be negligible, as the operation will also be recorded in the cudagraph, and the copy size is quite small.

I run the benchmark, and find the throughput diff is within run-to-run variations.

Before this PR:
Throughput: 18.57 requests/s, 9510.07 tokens/s

After this PR:
Throughput: 18.59 requests/s, 9516.03 tokens/s

**Comment by sfc-gh-hazhang** [comment]
I observe when enabling cuda_graph, the pp branch will use significantly more memory. This issue might be related?

**Comment by youkaichao** [comment]
> I observe when enabling cuda_graph, the pp branch will use significantly more memory. This issue might be related?

Yes it is possible. To cleverly use cudagraph with multiple sizes, we need some technique like sharing buffers, just as what we do for inputs, and what this PR does for outputs.

**Comment by youkaichao** [comment]
If this PR is merged, more cuda graphs can be captured at no additional memory cost, which leads to further questions:

- should we capture as many cuda graphs as we can (all batch sizes up until the maximum), so that we don't need padding anymore?
- should we reduce `_BATCH_SIZE_ALIGNMENT` from `8` to `4` or `2` , so that less computation is wasted on padding?
- should we align the batchsize used for tuning moe kernels with the cudagraph?

https://github.com/vllm-project/vllm/blob/fbdb7b3ee2c3f7b58841c12b52ed4c83f508babb/benchmarks/kernels/benchmark_mixtral_moe.py#L19-L22

cc @WoosukKwon for cudagraph
cc @pcmoritz for moe kernels

**Comment by youkaichao** [comment]
For llama2-7b:

before this PR:

```text
before capture graph with batch size 256 Used Memory: 19453.89 MB
after capture graph with batch size 256 Used Memory: 19464.02 MB
before capture graph with batch size 248 Used Memory: 19464.02 MB
after capture graph with batch size 248 Used Memory: 19465.95 MB
before capture graph with batch size 1 Used Memory: 19498.69 MB
after capture graph with batch size 1 Used Memory: 19498.70 MB
INFO 05-27 23:21:09 model_runner.py:920] Graph capturing finished in 6 secs.
```

the first graph takes 11MB memory, the second takes 2MB memory, and in total cudagraphs take 45MB memory.

after this PR:

```text
before capture graph with batch size 256 Used Memory: 19455.89 MB
after capture graph with batch size 256 Used Memory: 19464.02 MB
before capture graph with batch size 248 Used Memory: 19464.02 MB
after capture graph with batch size 248 Used Memory: 19464.02 MB
...
before capture graph with batch size 1 Used Memory: 19464.05 MB
after capture graph with batch size 1 Used Memory: 19464.05 MB
INFO 05-27 23:18:21 model_runner.py:927] Graph capturing finished in 8 secs.
```

the first graph takes 8MB memory, the second takes 0MB memory, and in total cudagraphs take 8MB memory. We can see the memory used after capturing all graphs (19464.05 MB) is almost the same as just capturing one graph previously (19464.02 MB).

**Comment by youkaichao** [comment]
throughput before this PR:

Throughput: 20.82 requests/s, 10658.60 tokens/s

throughput after this PR:

Throughput: 21.13 requests/s, 10818.10 tokens/s

**Comment by WoosukKwon** [comment]
@rkooo567 I can take this PR if you are ok with that. Just please let me know!

**Comment by rkooo567** [comment]
oh yeah please go ahead!

File: vllm/worker/model_runner.py [hunk 1]
```diff
@@ -871,6 +872,11 @@ def capture_model(self, kv_caches: List[torch.Tensor]) -> None:
         seq_lens = torch.ones(max_batch_size, dtype=torch.int32).cuda()
         block_tables = torch.from_numpy(self.graph_block_tables).cuda()
 
+        # Prepare buffer for outputs. These will be reused for all batch sizes.
+        hidden_states = torch.zeros(
+            (max_batch_size, self.model_config.get_output_size()),
+            dtype=self.model_config.dtype).cuda()
```
Comment (WoosukKwon) [inline vllm/worker/model_runner.py:878]: Why don't we reuse the output buffer of the first CUDA Graph (which is for the largest batch size)? Arguably, this will be more reliable than the current implementation.

**Review by WoosukKwon — COMMENTED** [review]
@youkaichao Thanks for the PR! This is a good finding. Please check my comments.

File: vllm/worker/model_runner.py (hunk 1 above)
Comment (youkaichao) [reply vllm/worker/model_runner.py:878]: That's doable. Although the code might look more ugly.

Comment (WoosukKwon) [reply vllm/worker/model_runner.py:878]: While it's debatable, I personally think it's more reliable than the current implementations which relies on HF config and `gc`. If the user has a custom model, the current implementation might break.

Comment (youkaichao) [reply vllm/worker/model_runner.py]: fixed in [dbd5248](https://github.com/vllm-project/vllm/pull/5074/commits/dbd5248db37e276c409f1c65eacde21398b2f8c9)

File: vllm/worker/model_runner.py [hunk 2]
```diff
@@ -993,13 +1001,21 @@ def capture(
...
                 **kwargs,
             )
+            if hidden_states is not None:
+                hidden_states.copy_(output_hidden_states)
+            else:
+                hidden_states = output_hidden_states
+            del output_hidden_states
+            # make sure `output_hidden_states` is deleted
+            # in the graph's memory pool
+            gc.collect()
```
Comment (WoosukKwon) [inline vllm/worker/model_runner.py:1018]: Why are these lines needed?

Comment (youkaichao) [reply vllm/worker/model_runner.py:1018]: see https://github.com/vllm-project/vllm/issues/5337 . `del x` is not guaranteed to release the resource of `x`.

Comment (WoosukKwon) [reply vllm/worker/model_runner.py:1018]: Oh my question was why we have to make sure `output_hidden_states` is deleted here.

Comment (youkaichao) [reply vllm/worker/model_runner.py:1018]: If `output_hidden_states` is not destructed in time, pytorch might think it is still in use, and the next cudagraph cannot reuse this part of memory.

Comment (WoosukKwon) [reply vllm/worker/model_runner.py:1018]: I see. Then why don't we call `gc.collect` in `capture_model` (at the upper layer)?

Comment (youkaichao) [reply vllm/worker/model_runner.py:1018]: we need to make sure `output_hidden_states` is destructed before we exit the `with torch.cuda.graph` context manager.

Comment (WoosukKwon) [reply vllm/worker/model_runner.py:1018]: Why do we need to? Could you elaborate more?

Comment (youkaichao) [reply vllm/worker/model_runner.py:1018]: iiuc, the memory sharing in pytorch cudagraph is:
- anything allocated and then destructed within `with torch.cuda.graph` can be reused (shared) for the next time. 
- Anything allocated but still alive outside of `with torch.cuda.graph` is persistent, and cannot be reused (shared) for the next time.

Comment (WoosukKwon) [reply vllm/worker/model_runner.py:1018]: I see. Thanks for the explanation!

**Review by WoosukKwon — APPROVED** [review]
LGTM!

**Comment by WoosukKwon** [comment]
@youkaichao Is the CI failure related to this PR?

**Comment by youkaichao** [comment]
> @youkaichao Is the CI failure related to this PR?

investigating.
